File: solutions/erlang/loops.erl
-module(loops).
-export([main/1]).

-define(SampleSize, 23).

% functions
random_day(_) -> rand:uniform(365).
random_sample(N) -> lists:map(fun(X) -> random_day(X) end, lists:seq(1, N)).
% identity(X) -> X. % cannot get (fun identity/1) to work
% ---------------------------------------------------------------------------
% performance notes
% recursive implementation is 4.5x slower than set implementation in escript
% but 50% faster when compiled
% stated another way: the recursive implementation gets >50x faster when compiled
% but the set implementation gets <10x faster
has_duplicates([]) -> false;
has_duplicates([X|XS]) -> case lists:member(X, XS) of true -> true; false -> has_duplicates(XS) end.
% has_duplicates(List) -> erlang:length(List) == sets:size(sets:from_list(List)).
% ---------------------------------------------------------------------------

main(Args) ->
  % data
  StartMicrosecond = os:system_time(),
  Iterations = list_to_integer(lists:nth(1, Args)),
  Data = lists:map(
    fun(_) -> has_duplicates(random_sample(?SampleSize)) end,
    lists:seq(1, Iterations)
  ),
  Duplicates = erlang:length(lists:filter(fun(X) -> X end, Data)),
  Percent = Duplicates / Iterations * 100,
  EndMicrosecond = os:system_time(),
  Seconds = (EndMicrosecond - StartMicrosecond) / 1000 / 1000 / 1000,
  % output
  io:fwrite("iterations: ~w~n", [Iterations]),
  io:fwrite("sample-size: ~w~n", [?SampleSize]),
  io:fwrite("percent: ~.2f~n", [Percent]),
  io:fwrite("seconds: ~.3f~n", [Seconds]).

File: solutions/clojure/loops.clj
; user input
(def iterations (read-string (first *command-line-args*)))

; constants
(def sample-size 23)

; functions
(defn random-day [_] (rand-int 365))
(defn create-sample [_] (map random-day (range sample-size)))
(defn vector-has-duplicates [x] (apply distinct? x))
(defn current-time-milliseconds [] (System/currentTimeMillis))

; data
(def start (current-time-milliseconds))
(def data (map create-sample (range iterations)))
(def duplicates (count (filter vector-has-duplicates data)))
(def percent (float (* (/ duplicates iterations) 100)))
(def finish (current-time-milliseconds))
(def seconds (float (/ (- finish start) 1000)))

; output
(println (str "iterations: " iterations))
(println (str "sample-size: " sample-size))
(printf "percent: %.2f%n" percent)
(printf "seconds: %.6f%n" seconds)
